design/StableRegions-20230209.md

```diff
@@ -0,0 +1,172 @@
+
+
+
+## Questions and answers
+
+### Q: What determines the 8MB non-empty region minimum?
+
+This size comes from wanting to grow a region by more than one page at
+a time (in terms of the way that the canister interacts with the
+system API, at least).  Rather than actually grow by a single page,
+the implementation grows by a "page block" (8MB) at a time.
+
+This choice means that there are 120 pages per page block, and that
+the maximum number of regions and region blocks are each relatively
+small (32k each).  Consequently, they can each be identified with a
+2-byte identifier, and we can pre-allocate tables to store certain
+relations about them, which is critical.
+
+### Q: Are 32767 regions enough?
+
+A: Permitting more than 32k regions may seem theoretically
+interesting, but is not practical given other parameters today that
+control the minimal footprint of a region (8MB) and dictate the
+maximum size of stable memory for a canister today (32GB).  With 32k
+regions at 8MB each, well over the maximum stable memory size is used
+(256GB compared to 32GB, the limit today)
+
+
+## Design details
+
+### API
+
+Internal region allocator operations:
+
+
+ - `initialize` -- called by the RTS, not by the Motoko developer.
+ - `rebuild` -- also called by the RTS, during an post-upgrade.
+
+
+User-facing region allocator operations:
+
+ - `region_new` -- create a dynamic region.
+ - `region_grow` -- grow region by a specified number of pages.
+ - `region_release`* -- release region and reuse region's page blocks.
+ - `region_load` -- read some data from the region.
+ - `region_store` -- store some data into the region.
+
+
+ - * The `region_release` operation is not part of the MVP, but supporting it
+   is important because it means we can transition quickly to an integration
+   with the ambient Motoko GC if we can support it.
+
+## Internal footprint
+
+The state of the allocator is stored in a combination of stable memory and heap memory.
+
+The stable memory state is sufficient to reconstitute the heap objects
+(`rebuild` operation, described in a subsection below).  That means
+that even if the heap is lost (like during an upgrade that fails) the
+stable memory state can fully describe the state parts that should be
+rebuilt into the heap when the upgrade succeeds.
+
+### stable memory fields
+
+ - total allocated buckets, `Nat16`, max value is `32768`.
+ - total allocated regions, `Nat16`, max value is `32767` (one region is reserved for "no region").
+ - bucket-region table (fixed size, about 131kb).
+ - region-object table (fixed size, about 393kb).
+
+### bucket-region table
+
+ - purpose: relate a bucket ID ("page block ID") to its region (if any) and its position in that region.
+ - 32768 entries (statically sized).
+ - 4 bytes per entry.
+ - entry type = `BucketRegion { region : Nat16, position : Nat16 }`
+ - the location of each entry gives its corresponding bucket ID.
+
+### region-object table
```

Comment:
The heap object pointers may not be necessary.  They were part of my original upgrade plan, where 
 - the `Region` value is stable by virtue of being represented as a simple ID (`Nat16`).
 - Each region operation uses that `id` to get at the heap object, and its vector of page block IDs.
 
 But now I realize that that's too costly, since it requires consulting stable memory to get that heap pointer.
 
 Instead, I have a new proposal now (for now just as replies to your comments).
 
 TL;DR -- the region heap object is stable, but its serialization is very simple (just a few fields, and no pointers) and its deserialization uses the preallocated tables to `rebuild` much like that description in the design file now.

Reply:
Cost was a concern, but it's not just a matter of cost. The references from stable memory to heap object would need to be treated as roots by the GC and updated when objects move, which complicates the GC. I think adding a new object tag is fairly easy, especially when it doesn't contain any heap object pointers (other logical pointers, i.e. offsets into stable memory, would just be bit patterns to the GC like other scalars).